{"sections":[],"sampleCodeDownload":{"action":{"identifier":"https:\/\/developer.apple.com\/wwdc20\/10687","isActive":true,"type":"reference","overridingTitle":"Watch Video (12 min)"},"kind":"sampleDownload"},"abstract":[{"text":"Put your test failures to work: Learn how to triage and diagnose uncaught issues in your app using the latest testing APIs in Xcode. We’ll show you how to help ease your testing workflow and put failures into context to help you deliver the best quality product.","type":"text"}],"metadata":{"modules":[{"name":"WWDC Notes"}],"roleHeading":"WWDC20","title":"Triage test failures with XCTIssue","role":"sampleCode"},"hierarchy":{"paths":[["doc:\/\/WWDCNotes\/documentation\/WWDCNotes","doc:\/\/WWDCNotes\/documentation\/WWDCNotes\/WWDC20"]]},"schemaVersion":{"patch":0,"major":0,"minor":3},"primaryContentSections":[{"content":[{"level":2,"text":"Overview","type":"heading","anchor":"Overview"},{"inlineContent":[{"type":"text","text":"Maintaining any active project requires investigating test failures, either locally or in Continuous Integration (CI). Investigating test failures requires answering the following questions:"}],"type":"paragraph"},{"items":[{"content":[{"type":"paragraph","inlineContent":[{"text":"What failed?","type":"text"}]}]},{"content":[{"type":"paragraph","inlineContent":[{"text":"How did it fail?","type":"text"}]}]},{"content":[{"type":"paragraph","inlineContent":[{"text":"Why did it fail?","type":"text"}]}]},{"content":[{"inlineContent":[{"type":"strong","inlineContent":[{"text":"Where","type":"text"}]},{"type":"text","text":" did it fail?"}],"type":"paragraph"}]}],"type":"unorderedList"},{"inlineContent":[{"text":"Xcode 12 makes answering these questions more efficient. This talk covers the following 4 topics:","type":"text"}],"type":"paragraph"},{"items":[{"content":[{"type":"paragraph","inlineContent":[{"text":"Swift errors in tests","type":"text"}]}]},{"content":[{"inlineContent":[{"text":"Rich failure objects","type":"text"}],"type":"paragraph"}]},{"content":[{"type":"paragraph","inlineContent":[{"type":"text","text":"Failure call stacks"}]}]},{"content":[{"type":"paragraph","inlineContent":[{"type":"text","text":"Advanced workflows"}]}]}],"type":"orderedList"},{"level":2,"text":"Presenting Test Failures in Xcode 12","type":"heading","anchor":"Presenting-Test-Failures-in-Xcode-12"},{"level":3,"text":"Issue Navigator","type":"heading","anchor":"Issue-Navigator"},{"inlineContent":[{"text":"This is an example of a test failure. Notice how the annotation is gray. This indicates that the failure happened underneath the indicated line. On the left, the issue navigator shows a call stack under the failed test.","type":"text"}],"type":"paragraph"},{"inlineContent":[{"identifier":"WWDC20-10687-gray_line","type":"image"}],"type":"paragraph"},{"inlineContent":[{"type":"text","text":"Clicking on a frame takes the source editor to that location. Now, the annotation is red because this is the actual source of the error."}],"type":"paragraph"},{"inlineContent":[{"type":"image","identifier":"WWDC20-10687-error_source"}],"type":"paragraph"},{"level":3,"text":"Test Report","type":"heading","anchor":"Test-Report"},{"inlineContent":[{"type":"text","text":"The Test Report is another way to explore test failures, and is especially useful when working with the results of a CI failure."}],"type":"paragraph"},{"inlineContent":[{"type":"text","text":"Here, we can see a list of failing tests, each with a call stack. When you hover over a stack frame, you can see two buttons to the right. The Jump button, indicated with an arrow, jumps to the source of the error. The second button, called the Assistant button, is new in Xcode 12. This does the same thing as the Jump button, but opens a secondary editor next to the Test Report so you don’t have to go back and forth. Clicking on a different stack frame changes the secondary editor immediately."}],"type":"paragraph"},{"inlineContent":[{"type":"image","identifier":"WWDC20-10687-assistant_button"}],"type":"paragraph"},{"level":2,"text":"1. Using Swift Errors in your tests","type":"heading","anchor":"1-Using-Swift-Errors-in-your-tests"},{"inlineContent":[{"text":"XCTest makes it possible for test functions to throw. When a test does throw, the error is used to formulate the error message.","type":"text"}],"type":"paragraph"},{"inlineContent":[{"type":"text","text":"This means you can replace boilerplate like this in your tests…"}],"type":"paragraph"},{"type":"codeListing","syntax":"swift","code":["func testExample() {","  do {","    try codeThatThrows()","  } catch {","    XCTFail(\"\\(error)\")","  }","}"]},{"inlineContent":[{"type":"text","text":"…with this:"}],"type":"paragraph"},{"type":"codeListing","syntax":"swift","code":["func testExample() throws {","  try codeThatThrows()","}"]},{"inlineContent":[{"type":"text","text":"However, until recently, these failures could not provide the source code location for thrown errors in tests. That has now been fixed with the Swift runtime improvements in the following platforms:"}],"type":"paragraph"},{"items":[{"content":[{"inlineContent":[{"type":"text","text":"iOS and tvOS 13.4"}],"type":"paragraph"}]},{"content":[{"type":"paragraph","inlineContent":[{"type":"text","text":"macOS 10.15.4"}]}]}],"type":"unorderedList"},{"level":3,"text":"Throwing from setUp and tearDown","type":"heading","anchor":"Throwing-from-setUp-and-tearDown"},{"inlineContent":[{"type":"text","text":"Similar improvements have also been brought to "},{"code":"setUp()","type":"codeVoice"},{"type":"text","text":" and "},{"code":"tearDown()","type":"codeVoice"},{"type":"text","text":"."}],"type":"paragraph"},{"inlineContent":[{"type":"image","identifier":"WWDC20-10687-throwing_setup_teardown"}],"type":"paragraph"},{"level":2,"text":"2. Rich Failure Objects","type":"heading","anchor":"2-Rich-Failure-Objects"},{"inlineContent":[{"text":"XCTest has always recorded test failures as 4 discrete values:","type":"text"}],"type":"paragraph"},{"items":[{"content":[{"type":"paragraph","inlineContent":[{"type":"text","text":"Failure message"}]}]},{"content":[{"inlineContent":[{"type":"text","text":"File path"}],"type":"paragraph"}]},{"content":[{"type":"paragraph","inlineContent":[{"text":"Line number where the failure was recorded","type":"text"}]}]},{"content":[{"inlineContent":[{"type":"text","text":"“Expected” flag to indicate if the failure was expected."}],"type":"paragraph"}]}],"type":"orderedList"},{"inlineContent":[{"type":"text","text":"These values were passed into the "},{"type":"codeVoice","code":"recordFailure"},{"type":"text","text":" API, which ensures that failures are logged and routed to Xcode for display:"}],"type":"paragraph"},{"type":"codeListing","syntax":"swift","code":["open class XCTestCase: XCTest {","","  open func recordFailure(","    withDescription: String,","    inFile: String,","    atLine: Int,","    expected: Bool","  )","","}"]},{"inlineContent":[{"text":"In Xcode 12, these values have been encapsulated into a new struct ","type":"text"},{"type":"codeVoice","code":"XCTIssue"},{"text":". This new type also adds:","type":"text"}],"type":"paragraph"},{"items":[{"content":[{"type":"paragraph","inlineContent":[{"type":"text","text":"Distinct types"}]}]},{"content":[{"inlineContent":[{"text":"Detailed description","type":"text"}],"type":"paragraph"}]},{"content":[{"type":"paragraph","inlineContent":[{"type":"text","text":"Associated error"}]}]},{"content":[{"type":"paragraph","inlineContent":[{"type":"text","text":"Attachments"}]}]}],"type":"unorderedList"},{"level":3,"text":"`XCTAttachment`","type":"heading","anchor":"XCTAttachment"},{"items":[{"content":[{"inlineContent":[{"text":"Can capture arbitrary data","type":"text"}],"type":"paragraph"}]},{"content":[{"type":"paragraph","inlineContent":[{"text":"Add this attachment to the test itself or an ","type":"text"},{"type":"codeVoice","code":"XCTActivity"}]}]},{"content":[{"inlineContent":[{"type":"codeVoice","code":"XCTIssue"},{"text":" supports attachments, which allows customized diagnostics for test failures.","type":"text"}],"type":"paragraph"}]}],"type":"unorderedList"},{"level":3,"text":"Recording `XCTIssue`s","type":"heading","anchor":"Recording-XCTIssues"},{"inlineContent":[{"text":"There is a new API on ","type":"text"},{"type":"codeVoice","code":"XCTestCase"},{"text":" for recording test failures, which is called by all types of asserts. The old ","type":"text"},{"type":"codeVoice","code":"recordFailure"},{"text":" API has been deprecated.","type":"text"}],"type":"paragraph"},{"type":"codeListing","syntax":"swift","code":["open class XCTestCase: XCTest {","","  open func record(_ issue: XCTIssue)","","}"]},{"level":3,"text":"Modifying issues","type":"heading","anchor":"Modifying-issues"},{"inlineContent":[{"text":"In Swift, issues are immutable or mutable depending on your use of ","type":"text"},{"code":"let","type":"codeVoice"},{"text":" or ","type":"text"},{"code":"var","type":"codeVoice"},{"text":":","type":"text"}],"type":"paragraph"},{"type":"codeListing","syntax":"swift","code":["let issue = XCTIssue(...)  \/\/ immutable","","var issue = XCTIssue(...)  \/\/ mutable"]},{"inlineContent":[{"text":"In Objective-C, there is a special subclass for mutable issues which conforms to ","type":"text"},{"code":"NSMutableCopy","type":"codeVoice"},{"text":":","type":"text"}],"type":"paragraph"},{"type":"codeListing","syntax":"objectivec","code":["XCTIssue *issue = [[XCTIssue alloc] init ...];  \/\/ immutable","","XCTMutableIssue *mutableIssue = [issue mutableCopy];  \/\/ mutable"]},{"level":2,"text":"3. Failures and Source Code","type":"heading","anchor":"3-Failures-and-Source-Code"},{"inlineContent":[{"type":"text","text":"One of the most important questions to answer about a test failure is “where did it happen?”."}],"type":"paragraph"},{"level":3,"text":"Before `XCTIssue`","type":"heading","anchor":"Before-XCTIssue"},{"inlineContent":[{"text":"Core test failure data always included file path and line number, which is great for simple tests. This is not enough for functions that are shared by more than one test.","type":"text"}],"type":"paragraph"},{"inlineContent":[{"text":"Here are two tests calling the same shared function:","type":"text"}],"type":"paragraph"},{"type":"codeListing","syntax":"swift","code":["func testThingA() throws {","  let thing = Thing(\"a\")","  try assertProperties(for: thing)","}","","func testThingB() throws {","  let thing = Thing(\"b\")","  try assertProperties(for: thing)","}","","\/\/ Shared function","func assertProperties(for thing: Thing) throws {","  XCTAssertNotNil(thing.name)","  XCTAssertNotNil(thing.order)","}"]},{"inlineContent":[{"type":"text","text":"Before, the test method would be marked as failing even though the failure occurred somewhere else. This has no further information to help the developer figure out "},{"type":"strong","inlineContent":[{"type":"text","text":"where"}]},{"type":"text","text":" the failure happened."}],"type":"paragraph"},{"inlineContent":[{"identifier":"WWDC20-10687-shared_failure","type":"image"}],"type":"paragraph"},{"inlineContent":[{"text":"This can be mitigated if the helper functions captures information from where it was invoked and explicitly uses that information in its assertion calls. However, if the helper has more than 1 assertion, there is now a different ambiguity: which assertion failed in the helper?","type":"text"}],"type":"paragraph"},{"inlineContent":[{"identifier":"WWDC20-10687-capture_helper","type":"image"}],"type":"paragraph"},{"level":3,"text":"Now with `XCTIssue`","type":"heading","anchor":"Now-with-XCTIssue"},{"inlineContent":[{"type":"strong","inlineContent":[{"code":"XCTIssue","type":"codeVoice"},{"text":" captures and symbolicates call stacks","type":"text"}]},{"text":" so the point of failure is more clear. Here, the same code now has a gray annotation to demonstrate that the failure happens lower in the call stack, and a red annotation where the failure actually occurs.","type":"text"}],"type":"paragraph"},{"inlineContent":[{"identifier":"WWDC20-10687-failure_with_xctissue","type":"image"}],"type":"paragraph"},{"inlineContent":[{"type":"text","text":"Most importantly, no extra effort was required to get this information."}],"type":"paragraph"},{"level":2,"text":"4. Advanced Workflows with new APIs","type":"heading","anchor":"4-Advanced-Workflows-with-new-APIs"},{"level":3,"text":"Advanced Workflow 1: Custom Assertions","type":"heading","anchor":"Advanced-Workflow-1-Custom-Assertions"},{"inlineContent":[{"text":"You can make your own assertions by creating ","type":"text"},{"type":"codeVoice","code":"XCTIssue"},{"text":" directly and then ","type":"text"},{"type":"codeVoice","code":"record(_ issue:)"},{"text":". In this example, the custom assertion validates some data and then includes the failure as an attachment to the issue that it records.","type":"text"}],"type":"paragraph"},{"type":"codeListing","syntax":"swift","code":["func assertSomething(","  about data: Data,","  file: StaticString = #filePath,","  line: UInt = #line",") {","  \/\/ Call out to custom validation function.","  guard !isValid(data) else { return }","","  \/\/ Create issue, declare with `var` for mutability.","  var issue = XCTIssue(type: .assertionFailure, compactDescription: \"Invalid data\")","","  \/\/ Attach the invalid data.","  issue.add(XCTAttachment(data: data))","","  \/\/ Capture the call site location as the point of failure.","  let location = XCTSourceCodeLocation(filepath: file, lineNumber: line)","  issue.sourceCodeContext = XCTSourceCodeContext(location: location)","","  \/\/ Record the issue.","  self.record(issue)","}"]},{"level":3,"text":"Advanced Workflow 2: Overriding `record(_ issue:)`","type":"heading","anchor":"Advanced-Workflow-2-Overriding-record-issue"},{"inlineContent":[{"text":"You can observe, suppress, or modify failures recorded in your test class. This method is the funnel point through which all issues pass, so overrides have total control over the output of the test class.","type":"text"}],"type":"paragraph"},{"level":4,"text":"Example 1: Overriding for Observation","type":"heading","anchor":"Example-1-Overriding-for-Observation"},{"type":"codeListing","syntax":"swift","code":["override func record(_ issue: XCTIssue) {","  \/\/ Observe, introspect, log, etc:","  if shouldLog(issue) {","    print(\"I just observed an issue!\")","  }","","  \/\/ Don't forget to call super!","  super.record(issue)","}"]},{"inlineContent":[{"text":"You can also use ","type":"text"},{"type":"codeVoice","code":"XCTestObservationCenter"},{"text":", but this approach is more useful if you only want to observe failures in one class.","type":"text"}],"type":"paragraph"},{"inlineContent":[{"type":"text","text":"If your override does "},{"type":"strong","inlineContent":[{"text":"not","type":"text"}]},{"type":"text","text":" call "},{"type":"codeVoice","code":"super"},{"type":"text","text":", you will have suppressed the issue. It will not continue along the recording chain, and nothing will be sent to Xcode."}],"type":"paragraph"},{"type":"codeListing","syntax":"swift","code":["override func record(_ issue: XCTIssue) {","  \/\/ If you don't want to record it, just return.","  if shouldSuppress(issue) {","    return","  }","","  \/\/ Otherwise, pass it to `super`.","  super.record(issue)","}"]},{"level":4,"text":"Example 2: Overriding for Modification","type":"heading","anchor":"Example-2-Overriding-for-Modification"},{"inlineContent":[{"text":"This is the most common reason for overriding because you can add attachments, which can be great diagnostic aids.","type":"text"}],"type":"paragraph"},{"type":"codeListing","syntax":"swift","code":["override func record(_ issue: XCTIssue) {","  \/\/ Redeclare using `var` to enable mutation.","  var issue = issue","","  \/\/ Add a simple attachment.","  issue.add(XCTAttachment(string: \"hello\"))","","  \/\/ Pass it to `super`.","  super.record(issue)","}"]},{"level":2,"text":"Wrap Up","type":"heading","anchor":"Wrap-Up"},{"items":[{"content":[{"type":"paragraph","inlineContent":[{"type":"text","text":"Triaging test failures is important!"}]}]},{"content":[{"inlineContent":[{"text":"Call stacks help answer “where?”","type":"text"}],"type":"paragraph"},{"type":"unorderedList","items":[{"content":[{"type":"paragraph","inlineContent":[{"type":"text","text":"This helps support more natural factoring of test code, such as code reuse and other best practices"}]}]}]}]},{"content":[{"type":"paragraph","inlineContent":[{"text":"Attachments improve “how?” and “why?”","type":"text"}]}]}],"type":"unorderedList"},{"level":2,"text":"Written By","type":"heading","anchor":"Written-By"},{"columns":[{"content":[{"type":"paragraph","inlineContent":[{"identifier":"skhillon","type":"image"}]}],"size":1},{"content":[{"level":3,"type":"heading","anchor":"Sarthak-Khillon","text":"Sarthak Khillon"},{"inlineContent":[{"type":"reference","isActive":true,"overridingTitleInlineContent":[{"text":"Contributed Notes","type":"text"}],"overridingTitle":"Contributed Notes","identifier":"doc:\/\/WWDCNotes\/documentation\/WWDCNotes\/skhillon"},{"type":"text","text":" "},{"type":"text","text":"|"},{"type":"text","text":" "},{"type":"reference","isActive":true,"identifier":"https:\/\/github.com\/skhillon"},{"type":"text","text":" "},{"type":"text","text":"|"},{"type":"text","text":" "},{"type":"reference","isActive":true,"identifier":"https:\/\/www.linkedin.com\/in\/sarthakkhillon\/"},{"type":"text","text":" "},{"type":"text","text":"|"},{"type":"text","text":" "},{"type":"reference","isActive":true,"identifier":"https:\/\/x.com\/SarthakKhillon"}],"type":"paragraph"}],"size":4}],"numberOfColumns":5,"type":"row"},{"inlineContent":[{"type":"text","text":"Missing anything? Corrections? "},{"type":"reference","identifier":"https:\/\/wwdcnotes.com\/documentation\/wwdcnotes\/contributing","isActive":true}],"type":"paragraph"},{"level":2,"text":"Related Sessions","type":"heading","anchor":"Related-Sessions"},{"style":"list","items":["doc:\/\/WWDCNotes\/documentation\/WWDCNotes\/WWDC20-10077-Eliminate-animation-hitches-with-XCTest","doc:\/\/WWDCNotes\/documentation\/WWDCNotes\/WWDC20-10091-Write-tests-to-fail","doc:\/\/WWDCNotes\/documentation\/WWDCNotes\/WWDC20-10164-XCTSkip-your-tests","doc:\/\/WWDCNotes\/documentation\/WWDCNotes\/WWDC20-10220-Handle-interruptions-and-alerts-in-UI-tests","doc:\/\/WWDCNotes\/documentation\/WWDCNotes\/WWDC20-10221-Get-your-test-results-faster"],"type":"links"},{"inlineContent":[{"inlineContent":[{"text":"Legal Notice","type":"text"}],"type":"strong"}],"type":"small"},{"inlineContent":[{"type":"text","text":"All content copyright © 2012 – 2024 Apple Inc. All rights reserved."},{"type":"text","text":" "},{"type":"text","text":"Swift, the Swift logo, Swift Playgrounds, Xcode, Instruments, Cocoa Touch, Touch ID, FaceID, iPhone, iPad, Safari, Apple Vision, Apple Watch, App Store, iPadOS, watchOS, visionOS, tvOS, Mac, and macOS are trademarks of Apple Inc., registered in the U.S. and other countries."},{"type":"text","text":" "},{"type":"text","text":"This website is not made by, affiliated with, nor endorsed by Apple."}],"type":"small"}],"kind":"content"}],"variants":[{"paths":["\/documentation\/wwdcnotes\/wwdc20-10687-triage-test-failures-with-xctissue"],"traits":[{"interfaceLanguage":"swift"}]}],"identifier":{"url":"doc:\/\/WWDCNotes\/documentation\/WWDCNotes\/WWDC20-10687-Triage-test-failures-with-XCTIssue","interfaceLanguage":"swift"},"kind":"article","references":{"doc://WWDCNotes/documentation/WWDCNotes/WWDC20-10220-Handle-interruptions-and-alerts-in-UI-tests":{"role":"sampleCode","url":"\/documentation\/wwdcnotes\/wwdc20-10220-handle-interruptions-and-alerts-in-ui-tests","type":"topic","title":"Handle interruptions and alerts in UI tests","kind":"article","abstract":[{"type":"text","text":"Learn how to anticipate potential interruptions to your app’s interface and build smart tests to identify them. UI interruptions often appear indeterminately, typically during onboarding or first launch, which can make them hard to track down. Learn how to understand interruptions, write stronger tests with UI interruption handlers, and manage expected alerts."}],"identifier":"doc:\/\/WWDCNotes\/documentation\/WWDCNotes\/WWDC20-10220-Handle-interruptions-and-alerts-in-UI-tests"},"skhillon":{"type":"image","alt":"Profile image of Sarthak Khillon","variants":[{"url":"\/images\/skhillon.jpeg","traits":["1x","light"]}],"identifier":"skhillon"},"https://x.com/SarthakKhillon":{"type":"link","titleInlineContent":[{"text":"X\/Twitter","type":"text"}],"title":"X\/Twitter","url":"https:\/\/x.com\/SarthakKhillon","identifier":"https:\/\/x.com\/SarthakKhillon"},"doc://WWDCNotes/documentation/WWDCNotes":{"role":"collection","title":"WWDC Notes","url":"\/documentation\/wwdcnotes","type":"topic","identifier":"doc:\/\/WWDCNotes\/documentation\/WWDCNotes","images":[{"identifier":"WWDCNotes.png","type":"icon"}],"abstract":[{"text":"Session notes shared by the community for the community.","type":"text"}],"kind":"symbol"},"WWDC20-10687-gray_line":{"type":"image","alt":null,"variants":[{"url":"\/images\/WWDC20-10687-gray_line.png","traits":["1x","light"]}],"identifier":"WWDC20-10687-gray_line"},"WWDCNotes.png":{"type":"image","alt":null,"variants":[{"url":"\/images\/WWDCNotes.png","traits":["1x","light"]}],"identifier":"WWDCNotes.png"},"doc://WWDCNotes/documentation/WWDCNotes/skhillon":{"url":"\/documentation\/wwdcnotes\/skhillon","type":"topic","abstract":[{"text":"No Bio on GitHub","type":"text"}],"identifier":"doc:\/\/WWDCNotes\/documentation\/WWDCNotes\/skhillon","kind":"article","title":"Sarthak Khillon (19 notes)","role":"sampleCode","images":[{"identifier":"skhillon.jpeg","type":"card"},{"identifier":"skhillon.jpeg","type":"icon"}]},"WWDC20-10687-failure_with_xctissue":{"type":"image","alt":null,"variants":[{"url":"\/images\/WWDC20-10687-failure_with_xctissue.png","traits":["1x","light"]}],"identifier":"WWDC20-10687-failure_with_xctissue"},"doc://WWDCNotes/documentation/WWDCNotes/WWDC20-10164-XCTSkip-your-tests":{"type":"topic","identifier":"doc:\/\/WWDCNotes\/documentation\/WWDCNotes\/WWDC20-10164-XCTSkip-your-tests","role":"sampleCode","url":"\/documentation\/wwdcnotes\/wwdc20-10164-xctskip-your-tests","title":"XCTSkip your tests","kind":"article","abstract":[{"text":"Get the test results that matter — and skip the ones that don’t. Discover how you can implement XCTSkip to conditionally avoid tests at runtime. We’ll take you through how to return this new test result and better document tests beyond pass and fail within your test bundle.","type":"text"}]},"WWDC20-10687-assistant_button":{"type":"image","alt":null,"variants":[{"url":"\/images\/WWDC20-10687-assistant_button.png","traits":["1x","light"]}],"identifier":"WWDC20-10687-assistant_button"},"doc://WWDCNotes/documentation/WWDCNotes/WWDC20":{"type":"topic","identifier":"doc:\/\/WWDCNotes\/documentation\/WWDCNotes\/WWDC20","abstract":[{"text":"Xcode 12, Swift 5.3, iOS 14, macOS 11 (Big Sur), tvOS 14, watchOS 7.","type":"text"},{"text":" ","type":"text"},{"type":"text","text":"New APIs: "},{"type":"codeVoice","code":"WidgetKit"},{"type":"text","text":", "},{"type":"codeVoice","code":"StoreKit Testing"},{"type":"text","text":", and more."}],"url":"\/documentation\/wwdcnotes\/wwdc20","kind":"article","role":"collectionGroup","title":"WWDC20","images":[{"type":"icon","identifier":"WWDC20-Icon.png"},{"type":"card","identifier":"WWDC20.jpeg"}]},"WWDC20-10687-capture_helper":{"type":"image","alt":null,"variants":[{"url":"\/images\/WWDC20-10687-capture_helper.png","traits":["1x","light"]}],"identifier":"WWDC20-10687-capture_helper"},"https://github.com/skhillon":{"type":"link","titleInlineContent":[{"text":"GitHub","type":"text"}],"title":"GitHub","url":"https:\/\/github.com\/skhillon","identifier":"https:\/\/github.com\/skhillon"},"doc://WWDCNotes/documentation/WWDCNotes/WWDC20-10077-Eliminate-animation-hitches-with-XCTest":{"title":"Eliminate animation hitches with XCTest","type":"topic","role":"sampleCode","identifier":"doc:\/\/WWDCNotes\/documentation\/WWDCNotes\/WWDC20-10077-Eliminate-animation-hitches-with-XCTest","url":"\/documentation\/wwdcnotes\/wwdc20-10077-eliminate-animation-hitches-with-xctest","abstract":[{"text":"Animations can dramatically enhance the user experience of your app, provide a sense of direct manipulation, and help people to better understand the results of their actions. Animation hitches can break that experience. Discover how to use XCTest to detect interruptions to smooth scrolling and animations, and learn how to catch regressions before they affect the people relying on your app.","type":"text"}],"kind":"article"},"WWDC20-10687-error_source":{"type":"image","alt":null,"variants":[{"url":"\/images\/WWDC20-10687-error_source.png","traits":["1x","light"]}],"identifier":"WWDC20-10687-error_source"},"WWDC20-10687-shared_failure":{"type":"image","alt":null,"variants":[{"url":"\/images\/WWDC20-10687-shared_failure.png","traits":["1x","light"]}],"identifier":"WWDC20-10687-shared_failure"},"WWDC20.jpeg":{"type":"image","alt":null,"variants":[{"url":"\/images\/WWDC20.jpeg","traits":["1x","light"]}],"identifier":"WWDC20.jpeg"},"https://www.linkedin.com/in/sarthakkhillon/":{"type":"link","titleInlineContent":[{"text":"Blog","type":"text"}],"title":"Blog","url":"https:\/\/www.linkedin.com\/in\/sarthakkhillon\/","identifier":"https:\/\/www.linkedin.com\/in\/sarthakkhillon\/"},"https://wwdcnotes.com/documentation/wwdcnotes/contributing":{"type":"link","titleInlineContent":[{"text":"Contributions are welcome!","type":"text"}],"title":"Contributions are welcome!","url":"https:\/\/wwdcnotes.com\/documentation\/wwdcnotes\/contributing","identifier":"https:\/\/wwdcnotes.com\/documentation\/wwdcnotes\/contributing"},"skhillon.jpeg":{"type":"image","alt":null,"variants":[{"url":"\/images\/skhillon.jpeg","traits":["1x","light"]}],"identifier":"skhillon.jpeg"},"doc://WWDCNotes/documentation/WWDCNotes/WWDC20-10091-Write-tests-to-fail":{"url":"\/documentation\/wwdcnotes\/wwdc20-10091-write-tests-to-fail","role":"sampleCode","type":"topic","kind":"article","title":"Write tests to fail","identifier":"doc:\/\/WWDCNotes\/documentation\/WWDCNotes\/WWDC20-10091-Write-tests-to-fail","abstract":[{"type":"text","text":"Plan for failure: Design great tests to help you find and diagnose even the toughest bugs. Learn how to improve your automated tests with XCTest to find hidden issues in even the best code. We’ll explain how to prepare your tests for failure to make triaging issues easier, letting you solve interface issues and deliver fixes quickly."}]},"WWDC20-Icon.png":{"type":"image","alt":null,"variants":[{"url":"\/images\/WWDC20-Icon.png","traits":["1x","light"]}],"identifier":"WWDC20-Icon.png"},"WWDC20-10687-throwing_setup_teardown":{"type":"image","alt":null,"variants":[{"url":"\/images\/WWDC20-10687-throwing_setup_teardown.png","traits":["1x","light"]}],"identifier":"WWDC20-10687-throwing_setup_teardown"},"https://developer.apple.com/wwdc20/10687":{"type":"download","url":"https:\/\/developer.apple.com\/wwdc20\/10687","identifier":"https:\/\/developer.apple.com\/wwdc20\/10687","checksum":null},"doc://WWDCNotes/documentation/WWDCNotes/WWDC20-10221-Get-your-test-results-faster":{"url":"\/documentation\/wwdcnotes\/wwdc20-10221-get-your-test-results-faster","title":"Get your test results faster","role":"sampleCode","abstract":[{"text":"Improve your testing suite to speed up your feedback loop and get fixes in faster. Learn more about the latest improvements to testing in Xcode, including how to leverage test plans, Xcodebuild updates, and APIs to eliminate never-ending and badly-behaved tests. We’ll explore Test Timeouts and Execution Time Allowances in XCTest, examine device parallelization, and detail recommended practices for balancing performance with clear fault localization.","type":"text"}],"kind":"article","identifier":"doc:\/\/WWDCNotes\/documentation\/WWDCNotes\/WWDC20-10221-Get-your-test-results-faster","type":"topic"}}}